src/bgen/Generator.cs

```diff
@@ -2973,20 +2973,26 @@ void GenerateInvoke (bool stret, bool supercall, MethodInfo mi, MemberInformatio
 
 			if (minfo.is_static)
 				print ("{0}{1}{2} (class_ptr, {5}{6}){7};",
-					   returns ? (assign_to_temp ? "ret = " : "return ") : "",
+					   returns ? "ret = " : "",
 					   cast_a, sig, target_name,
 					   "/*unusued3*/", //supercall ? "Super" : "",
 					   selector_field, args, cast_b);
 			else
 				print ("{0}{1}{2} ({3}{4}, {5}{6}){7};",
-					   returns ? (assign_to_temp ? "ret = " : "return ") : "",
+					   returns ? "ret = " : "",
 					   cast_a, sig, target_name,
 					   handle,
 					   selector_field, args, cast_b);
 
 			if (postproc.Length > 0)
 				print (postproc.ToString ());
 		}
+
+		if (target_name == "This") {
```

Comment:
IMHO this should not be condition on `"This"`, but instead of whether the method is a an actual extension method. So further above:


```cs
var useExtensionMethod = (category_type is null && !minfo.is_extension_method && !minfo.is_protocol_implementation_method);
string target_name = useExtensionMethod ? "this" : "This";
```

and then here:

```suggestion
		if (useExtensionMethod) {
```

Reply:
Sure, but will be the opposite, useInstanceMethod:
```csharp
var useExtensionMethod = (category_type is null && !minfo.is_extension_method && !minfo.is_protocol_implementation_method);
```
Is ensuring that we are not a category etc.. We could move it to an or, but I don't like to change bool operations just because.

Reply:
ah, yes, flip the logic:

```cs
var useExtensionMethod = !(category_type is null && !minfo.is_extension_method && !minfo.is_protocol_implementation_method);
string target_name = useExtensionMethod ? "This" : "this";
```

boolean logic is hard!